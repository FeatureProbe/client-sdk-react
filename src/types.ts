import { FeatureProbe, FPConfig, FPDetail } from 'featureprobe-client-sdk-js';

export interface IContext {
  client?: FeatureProbe,
  toggles?: { [key: string]: FPDetail } | undefined,
}

export interface IProviderProps {
  config?: FPConfig,
  FPClient?: FeatureProbe,
}