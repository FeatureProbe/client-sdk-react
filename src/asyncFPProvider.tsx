import React, { ReactNode } from 'react'
import { FeatureProbe, FPDetail, FPConfig } from 'featureprobe-client-sdk-js';
import FPContext from './context';

export default async function AsyncFPProvider(config: FPConfig) {
  let toggles: { [key: string]: FPDetail; } | undefined;
  const client = new FeatureProbe(config);

  client.on('update', () => {
    toggles = client.allToggles();
  });

  client.start();

  await client.waitUntilReady();
  toggles = client.allToggles();

  const FPProvider = ({ children }: { children: ReactNode }) => {
    return (
      <FPContext.Provider 
        value={{
          client: client,
          toggles
        }}
      >
        {children}
      </FPContext.Provider>
    );
  };

  return FPProvider;
}
