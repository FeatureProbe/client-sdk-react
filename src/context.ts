import { createContext } from 'react';
import { IContext } from './types';

const FPContext = createContext<IContext>(null as never);

export default FPContext;
