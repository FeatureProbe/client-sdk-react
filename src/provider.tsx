import { FeatureProbe, FPDetail } from 'featureprobe-client-sdk-js';
import FPContext from './context';
import { IProviderProps } from './types';

import React, {
  FC,
  PropsWithChildren,
  useEffect,
  useState,
  useRef
} from 'react';

const FPProvider: FC<PropsWithChildren<IProviderProps>> = ({
  config,
  children,
  FPClient
}) => {
  let client = useRef<FeatureProbe>();
  const [toggles, saveToggles] = useState<{ [key: string]: FPDetail; } | undefined>();

  if (!config && !FPClient) {
    throw new Error('config or FPClient is required');
  }

  if (FPClient) {
    client.current = FPClient;
  }

  if (!client.current && config) {
    client.current = new FeatureProbe(config);
  }

  useEffect(() => {
    client.current?.stop();
    client.current?.start();

    client.current?.on('update', () => {
      saveToggles(client.current?.allToggles());
    }); 

    return function cleanup() {
      client.current?.off('update', () => {});
      client.current?.stop();
    };
  }, []);

  return (
    <FPContext.Provider 
      value={{
        client: client.current,
        toggles
      }}
    >
      {children}
    </FPContext.Provider>
  );
};

export default FPProvider;
