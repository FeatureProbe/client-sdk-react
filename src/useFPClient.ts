import { useContext } from 'react';
import FPContext from './context';

const useFPClient = () => {
  const { client } = useContext(FPContext);
  return client;
};

export default useFPClient;
