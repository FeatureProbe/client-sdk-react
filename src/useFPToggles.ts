import { useContext } from 'react';
import FPContext from './context';

const useFPToggles = () => {
  const { toggles } = useContext(FPContext);
  return toggles;
};

export default useFPToggles;
