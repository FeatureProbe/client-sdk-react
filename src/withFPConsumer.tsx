import React, { ComponentType } from 'react';
import FPContext from './context';
import { IContext } from './types';

export function withFPConsumer<C>(
  WrappedComponent: ComponentType<C & IContext>
) {
  return (props: C) => (
    <FPContext.Consumer>
      {
        ({ client, toggles }) => (
          <WrappedComponent client={client} toggles={toggles} {...props} />
        )
      }
    </FPContext.Consumer>
  )
}
