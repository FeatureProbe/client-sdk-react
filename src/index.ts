import FPProvider from './provider';
import useFPClient from './useFPClient';
import useFPToggles from './useFPToggles';
import { withFPConsumer } from './withFPConsumer';
import AsyncFPProvider from './asyncFPProvider';
import { getPlatform, initializePlatform } from 'featureprobe-client-sdk-js';
import { version } from '../package.json';

export { FeatureProbe, FPUser } from 'featureprobe-client-sdk-js';
export type { FPDetail, FPConfig, FPStorageProvider } from 'featureprobe-client-sdk-js';

initializePlatform({
  platform: {
    ...getPlatform(),
    UA: 'REACT/' + version,
  }
});

export {
  FPProvider,
  AsyncFPProvider,
  useFPClient,
  useFPToggles,
  withFPConsumer,
}