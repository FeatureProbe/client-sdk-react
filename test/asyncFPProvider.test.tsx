import React from 'react';
import { render } from '@testing-library/react';
import { FPConfig, FPUser } from 'featureprobe-client-sdk-js';
import '@testing-library/jest-dom';
import { FetchMock } from "jest-fetch-mock";
import * as data from "./fixtures/toggles.json";
import { AsyncFPProvider, withFPConsumer } from '../src';

const givenConfig: FPConfig = {
  clientSdkKey: 'my-app',
  user: new FPUser(),
  remoteUrl: 'http://127.0.0.1:4007',
  enableAutoReporting: false,
};

const _fetch = fetch as FetchMock;

test('should start SDK with AsyncFPProvider successfully', async () => {
  _fetch.mockResponseOnce(JSON.stringify(data));
  const providerProps = {
    config: givenConfig,
  };

  const App = withFPConsumer((props) => {
    const { toggles } = props;
    return (
      <span>Received: {JSON.stringify(toggles)}</span>
    );
  });

  const FPProvider = await AsyncFPProvider(providerProps.config);
  const { container } = render(
    <FPProvider>
      <App />
    </FPProvider>
  );

  expect(container).toMatchSnapshot();
});





