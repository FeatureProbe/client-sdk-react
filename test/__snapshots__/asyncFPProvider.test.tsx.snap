// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should start SDK with AsyncFPProvider successfully 1`] = `
<div>
  <span>
    Received: 
    {"json_toggle":{"value":{"v":"v1","variation_0":"c2"},"ruleIndex":0,"version":1,"reason":"rule 0 "},"multi_condition_toggle":{"value":{"variation_0":""},"ruleIndex":0,"version":1,"reason":"rule 0 "},"number_toggle":{"value":1,"ruleIndex":0,"version":1,"reason":"rule 0 "},"disabled_toggle":{"value":{"disabled_key":"disabled_value"},"ruleIndex":null,"version":1,"reason":"disabled"},"string_toggle":{"value":"1","ruleIndex":0,"version":1,"reason":"rule 0 "},"bool_toggle":{"value":true,"ruleIndex":0,"version":1,"reason":"rule 0 "},"default":{"json_toggle":{"value":{"v":"v1","variation_0":"c2"},"ruleIndex":0,"version":1,"reason":"rule 0 "},"multi_condition_toggle":{"value":{"variation_0":""},"ruleIndex":0,"version":1,"reason":"rule 0 "},"number_toggle":{"value":1,"ruleIndex":0,"version":1,"reason":"rule 0 "},"disabled_toggle":{"value":{"disabled_key":"disabled_value"},"ruleIndex":null,"version":1,"reason":"disabled"},"string_toggle":{"value":"1","ruleIndex":0,"version":1,"reason":"rule 0 "},"bool_toggle":{"value":true,"ruleIndex":0,"version":1,"reason":"rule 0 "}}}
  </span>
</div>
`;
